src/otaclient/metrics.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2022 TIER IV, INC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import json
+import logging
+from dataclasses import asdict, dataclass
+
+from _otaclient_version import __version__
+
+from otaclient._logging import LogType
+from otaclient.configs.cfg import (
+    ecu_info,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class OTAMetricsData:
+    """
+    Dataclass for storing metrics data.
+    These data are collected during the OTA update process, converted to JSON, and expected to be used in subscription filters.
+    Thus, flatten structure is used for easy access.
+    """
+
+    # Date
+    initializing_start_timestamp: int = 0
+    processing_metadata_start_timestamp: int = 0
+    delta_calculation_start_timestamp: int = 0
+    download_start_timestamp: int = 0
+    apply_update_start_timestamp: int = 0
+    post_update_start_timestamp: int = 0
+    finalizing_update_start_timestamp: int = 0
+
+    # ECU and Firmware
+    ecu_id: str = ecu_info.ecu_id
+    session_id: str = ""
+    current_firmware_version: str = ""
+    target_firmware_version: str = ""
+
+    # OTA Client
+    otaclient_version: str = __version__
+
+    # Status
+    failure_type: str = ""
+    failure_reason: str = ""
+    failed_at_phase: str = ""
+
+    # Metrics
+    ota_image_total_files_size: int = 0
+    ota_image_total_regulars_num: int = 0
+    ota_image_total_directories_num: int = 0
+    ota_image_total_symlinks_num: int = 0
+    delta_download_files_num: int = 0
+    delta_download_files_size: int = 0
+    downloaded_bytes: int = 0
+    downloaded_errors: int = 0
+
+    def __init__(self):
+        # this variable will not be included in data fields
+        self._already_published = False
+
```

Comment:
Should be `__post_init__`, not `__init__` :pray: .
Manual defined `__init__` will conflict with the dataclass generated `__init__`.

Reply:
Good catch! Fixed.